### PR TITLE
Feature 18367 configurable metric formatting

### LINKED ISSUE #18367: Customizable Metric Formatting for Rich Progress Bar
### Description & Motivation

Allow users to customize the formatting of metric values in the Rich Progress Bar.

#### Background:
Currently, PyTorch Lightning's Rich Progress Bar formats all float values using the approach: `round(value, 3)`. While this approach works well for many common scenarios, it becomes ineffective for certain edge cases where the loss values are either too large or too small, rendering the displayed values not informative or potentially misleading.

#### Proposal:
Introduce an option to specify a custom formatting string for metrics displayed in the Rich Progress Bar. This would provide flexibility and cater to a broader range of use cases.

#### Benefits:
* Versatility: Users working with metrics in atypical ranges would be able to display them more effectively.
* Clarity: A custom format might help users present metric values in a more understandable or meaningful manner, aiding quick insights during training.
* Backward Compatibility: By keeping the default as round(value, 3), existing users won't see any changes unless they opt to use the new feature.

#### Suggested Implementation:

1) Add an optional parameter, say metric_format, to the Rich Progress Bar's initialization. This parameter would accept Python string format specifications.
2) Use this format string to display metric values if provided; otherwise, default to the existing behavior.

Optionally, I would like to see another optional flag named scientific_notation. When set to True, it should override the metric_format value and set it to ".3e" for displaying metrics in scientific notation format. If the user provides both the metric_format and sets scientific_notation to True, a warning should be issued informing the user that the scientific_notation flag will take precedence. This flag would increase usability because the scientific notation would be the most common custom formatting that a user needs to specify.

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by quintenroets**
## What does this PR do?

Fixes #18367 

Blocked by https://github.com/Lightning-AI/lightning/pull/18372

This PR makes the string formatting of metric values configurable.
This allows for tracking losses in non-standard scenarios (e.g losses in the range 1e-7). 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]
Would it be possible to provide a test as well?

File: src/lightning/pytorch/callbacks/progress/rich_progress.py [hunk 1]
```diff
@@ -173,8 +180,10 @@ def render(self, task: "Task") -> Text:
             return Text(text, justify="left", style=self._style)
 
         def _generate_metrics_texts(self) -> Generator[str, None, None]:
-            for k, v in self._metrics.items():
-                yield f"{k}: {round(v, 3) if isinstance(v, float) else v}"
+            for name, value in self._metrics.items():
+                if not isinstance(value, str):
+                    value = round(value, 3) if self._metrics_format is None else f"{value:{self._metrics_format}}"
```
Comment (carmocca) [inline src/lightning/pytorch/callbacks/progress/rich_progress.py:185]: `metrics_format` could have `.3f` as default which should be equivalent to `round(, 3)`:

Comment (quintenroets) [reply src/lightning/pytorch/callbacks/progress/rich_progress.py:185]: I definitely agree. It would not entirely be backward compatible for metrics with lower precision. That's why I didn't propose it originally.

e.g.
100.1 becomes
* round(, 3) => 100.1
* .3f => 100.100

In my opinion, the .3f option is actually better than the current behavior because it results in consistent metrics lengths, making the progress bar values jump around less, and making the metrics more readable. (For example, if 10.111 changes to 10.1, this introduces a shift in all other metric values)

Comment (carmocca) [reply src/lightning/pytorch/callbacks/progress/rich_progress.py:185]: Good point. I don't think we need to be this strict about backward compatibility in terms of how the data is presented on the progress bar. Especially if the change is it's for a better UI

**Comment by quintenroets** [comment]
> Thanks for the contribution 🚀
> 
> Would it be possible to provide a test as well?

Of course, I added two tests at https://github.com/Lightning-AI/lightning/pull/18483

File: src/lightning/pytorch/callbacks/progress/rich_progress.py (hunk 1 above)
Comment (quintenroets) [reply src/lightning/pytorch/callbacks/progress/rich_progress.py:185]: Okay, I agree. I created a PR for this at https://github.com/Lightning-AI/lightning/pull/18483
